File: a-feast-for-cats/feast4cats.py

# https://open.kattis.com/problems/cats 

# optimize https://www.geeksforgeeks.org/dsa/introduction-to-disjoint-set-data-structure-or-union-find-algorithm/

from functools import cmp_to_key
import time

def comparator(a,b):
    return a[2] - b[2]


class Edge:
  def __init__(self, edgeFrom, edgeTo, weight):
    self.edgeFrom = edgeFrom
    self.edgeTo = edgeTo
    self.weight = weight

def sortFunc(e):
   return e[2]

class DSU:
  def __init__(self, n):
    self.parent = list(range(n))
    self.rank = [1] * n

  def find(self, i):
    if self.parent[i] != i:
      self.parent[i] = self.find(self.parent[i])
    return self.parent[i]

  def union(self, x, y):
    s1 = self.find(x)
    s2 = self.find(y)
    if s1 != s2:
      if self.rank[s1] < self.rank[s2]:
        self.parent[s1] = s2
      elif self.rank[s1] > self.rank[s2]:
        self.parent[s2] = s1
      else:
        self.parent[s2] = s1
        self.rank[s1] += 1

def kruskal(f):
  # milkAndCats = input()
  milkAndCats = f.readline()
  milkAndCats = milkAndCats.split()

  milk = int(milkAndCats[0])
  cats = int(milkAndCats[1])

  edgeList = []

  edges = int((cats * (cats - 1)) / 2)

  for i in range(edges) :
      edge = f.readline() # input()
      edge = edge.split()
      edgeList.append([int(edge[0]), int(edge[1]), int(edge[2])])
      edgeList.append([int(edge[1]), int(edge[0]), int(edge[2])])

  edgeList.sort(key=cmp_to_key(comparator))

  dsu = DSU(cats)
  cost = 0
  count = 0

  for edgeFrom, edgeTo, weight in edgeList:
    if(dsu.find(edgeFrom) != dsu.find(edgeTo)):
      dsu.union(edgeFrom, edgeTo)
      cost += weight
      count += 1
      if count == cats - 1:
        break

  if(milk >= cost):
    print('yes milk: {}, cost: {}'.format(milk, cost))
  else: 
    print('no milk: {}, cost: {}'.format(milk, cost))


if __name__ == '__main__':
  with open("graph-testdata.txt") as f: 
    start_time = time.time()
    testCases = f.readline()
    testCases = int(testCases)
    i = 0
    while i < testCases:
      kruskal(f)
      i = i + 1 
  print("time: {}".format(time.time() - start_time))

File: a-feast-for-cats/feast4cats.go
package main

import (
	"fmt"
	"sort"
	"container/heap"
)

func kruskals() string {
	var milk, cats int
	fmt.Scanf("%d %d%n", &milk, &cats)
	numOfEdges := int((cats * (cats - 1)) / 2)

	if cats > milk {
		for i := 0; i < numOfEdges; i++ {
			fmt.Scanf("%d %d")
		}
		return "no\n"
	}

	edges := make([]Edge, numOfEdges)
	j := 0

	adjList := make([][]Edge, numOfEdges)
	visited := make([]bool, cats)

	for i := range numOfEdges {
		adjList[i] = make([]Edge, cats - 1)
		var edge Edge
		var from, to, weight int
		fmt.Scanf("%d %d %d", &from, &to, &weight)
		adjList[from] = append(edges[from], Edge{from, to, weight})
		adjList[to] = append(edges[to], Edge{to, from, weight})

	}



	cost := 0
	queue := &EdgeHeap{}
	heap.Init(queue)
	heap.Push()

	while()
	for _, edge := range edges {
		if dsu.find(edge.from) != dsu.find(edge.to) {
			dsu.union(edge.from, edge.to)
			cost += edge.weight
			count++
			if count == cats - 1 {
				break
			}
			if cost > milk {
				return "no\n"
			}
		}
	}

	return "yes\n"
}

func main() {
	var testCases int

	fmt.Scanf("%d%n", &testCases)
	
	var out string

	for i:= 0; i < testCases; i++ {
		out += kruskals()
	}
	fmt.Print(out)
}

type Edge struct {
	from int
	to int
	weight int
}

// An EdgeHeap is a min-heap of ints.
type EdgeHeap []Edge

func (h EdgeHeap) Len() int           { return len(h) }
func (h EdgeHeap) Less(i, j int) bool { return h[i].weight < h[j].weight }
func (h EdgeHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *EdgeHeap) Push(x any) {
	// Push and Pop use pointer receivers because they modify the slice's length,
	// not just its contents.
	*h = append(*h, x.(Edge))
}

func (h *EdgeHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}
